odoo/addons/base/transpiler/sourcemap_generator.py

```diff
@@ -0,0 +1,177 @@
+            ls 1 <=> lg previous_genererated_line + 1
+
+            Step 2:
+            ls 1 <=> lg previous_genererated_line + nb_skip_line_output + 1
+            ls 2 <=> lg previous_genererated_line + nb_skip_line_output + 2
+            ...
+            ls nb_ls <=> lg previous_genererated_line + nb_skip_line_output + nb_ls
+
+
+        :param source_name: name of the source to add
+        :param source_content: content of the source to add
+        :param previous_genererated_line: Line where we start to map the new source
+        :param nb_skip_line_output: Number of lines to pass in the generated file before starting mapping line by line
+
+        :return number of lines added
+        """
+        source_line_count = len(source_content.split("\n"))
+
+        if (source_name not in self._sources):
+            self._sources[source_name] = len(self._sources)
+
+        self._sourcesContents[source_name] = source_content
+        if nb_skip_line_output > 0:
+            # adds a mapping between the first line of the source
+            # and the first line of the corresponding code in the generated file.
+            self._mappings.append({
+                "generatedLine": previous_genererated_line + 1,
+                "originalLine": 1,
+                "source": source_name,
+            })
+        for i in range(1, source_line_count + 1):
+            self._mappings.append({
+                "generatedLine": previous_genererated_line + i + nb_skip_line_output,
+                "originalLine": i,
+                "source": source_name,
+            })
+        return source_line_count + nb_skip_line_output
+
+
+def memorize(fn):
+    cache = {}
+
+    def fn_memoize(args):
+        if args not in cache:
+            cache[args] = fn(args)
+        return cache[args]
+    return fn_memoize
+
+
+def base64vlq():
+    """Encode Base64 VLQ encoded sequences
+    https://gist.github.com/mjpieters/86b0d152bb51d5f5979346d11005588b
+    Base64 VLQ is used in source maps.
+    VLQ values consist of 6 bits (matching the 64 characters of the Base64
+    alphabet), with the most significant bit a *continuation* flag. If the
+    flag is set, then the next character in the input is part of the same
+    integer value. Multiple VLQ character sequences so form an unbounded
+    integer value, in little-endian order.
+    The *first* VLQ value consists of a continuation flag, 4 bits for the
+    value, and the last bit the *sign* of the integer:
+    +-----+-----+-----+-----+-----+-----+
+    |  c  |  b3 |  b2 |  b1 |  b0 |  s  |
+    +-----+-----+-----+-----+-----+-----+
+    while subsequent VLQ characters contain 5 bits of value:
+    +-----+-----+-----+-----+-----+-----+
+    |  c  |  b4 |  b3 |  b2 |  b1 |  b0 |
+    +-----+-----+-----+-----+-----+-----+
+    For source maps, Base64 VLQ sequences can contain 1, 4 or 5 elements.
+    """
+
+    _b64chars = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
```

Comment:
question: would it be better to define these constants out of this function in the scope of this file and then simply define a function which uses them, instead of capturing them in the closure like this?

Reply:
Yes. And if performance is a concern it would be best to then set them as default values on a kwarg, this woudl also avoid having to write e.g.

        shiftsize, flag, mask = _shiftsize, _flag, _mask

(well it'd be written in the params instead, I guess)